databuilder/query_model.py

```diff
@@ -378,47 +378,63 @@ def validate_input_domains(node):
+    #   to combine `Table A -> Filter B -> Filter C` and a series derived from
+    #   `Table A -> Filter B`. But it's not legal to combine `Table A -> Filter D` and a
+    #   series derived from `Table A -> Filter E`. Combination of this nature takes place
+    #   with the Filter and Sort operations.
     #
-    #        foobar_events = events.filter(foo_events.date.isin(bar_dates))
+    #   That means that filters and sorts can be expressed tersely in ehrQL, like this:
+    #
+    #       events.filter(events.code == "foo").filter(events.date >= start_date)
     #
     # We enforce this by modelling domains as a hierarchy. At the root is the patient
     # domain, each table has an unique domain descending from this, and each filter
-    # operation creates a new domain descending from the domain of its source. A valid
-    # operation is one where the domains of its inputs are "nested" in the sense that
-    # each domain is an ancestor or descendant of every other.
+    # operation creates a new domain descending from the domain of its source.
     #
-    # We use sets to implement this hierarchy: we start with the patient domain as a set
-    # containing a single unique value, and we create descendent domains by forming the
-    # union of the parent domain with a new unqiue value. The `issubset()` relation then
-    # gives us the "is domain ancestor" semantics we want.
+    # A valid series combination operation is one where the non-patient domains are
+    # identical.  A valid filter or sort operation is one where the domains of its series
+    # input's domain is an ancestor of its frame input's domain.
     #
-    # Another way of expressing the "nested" condition above is that the input domains
-    # must be totally ordered by ancestory. We can check this by sorting them (sets
-    # naturally sort by subset-hood in Python) and then confirming that the ordering is
-    # total by testing that each adjacent pair has the required subset relation.
-    domains = sorted(get_input_domains(node))
-    for a, b in zip(domains, domains[1:]):
-        if not a.issubset(b):
+    # We use sets to implement this hierarchy: we start with the patient domain as a set
+    # containing a single unique value, and we create descendant domains by forming the
+    # union of the parent domain with a new unique value.
+
+    if isinstance(node, Filter) or isinstance(node, Sort):
+        source_domain = get_domain(node.source)
+
+        if isinstance(node, Filter):
+            series = node.condition
+        else:
+            series = node.sort_by
+        series_domain = get_domain(series)
+
+        if not series_domain.issubset(source_domain):
+            raise DomainMismatchError(
+                f"Attempt to combine series with domain:\n{series_domain}"
+                f"\nWith frame with domain:\n{source_domain}"
+                f"\nIn node:\n{node}"
+            )
```

Comment:
I think this is fine but fwiw, the approach I took when I sketched this out earlier was to use `@singledispatch` on `validate_input_domains()`  and then have a single function which works with both `Filter` and `Sort`:
```py
@validate_input_domains.register(Filter)
@validate_input_domains.register(Sort)
def validate_inputs_domains_for_frame_and_series_operations(node):
    frame, series = get_input_nodes(node)
    assert isinstance(frame, Frame)
    assert isinstance(series, Series)
    frame_domain = get_domain(frame)
    series_domain = get_domain(series)
    if not series_domain.issubset(frame_domain):
        raise DomainMismatchError(
            f"Attempt to combine series with domain:\n{series_domain}"
            f"\nWith frame with domain:\n{frame_domain}"
            f"\nIn node:\n{node}"
        )
```

I think `@singledispatch` probably isn't helpful here, but I quite like not caring what the argument name of the series is as I feel it makes the intent clearer.